src/validator/nsec.rs

```diff
@@ -0,0 +1,1034 @@
+            } else if types.contains(Rtype::NS) && !types.contains(Rtype::SOA)
+            {
+                // totest, non-DS Rtype and NSEC from parent
+                // This is an NSEC record from the parent. Complain.
+                let ede = make_ede(
+                    ExtendedErrorCode::DNSSEC_BOGUS,
+                    "NSEC from parent for non-DS rtype",
+                );
+                return (NsecState::Nothing, ede);
+            }
+
+            // Anything else is a secure intermediate node.
+            return (NsecState::NoData, None);
+        }
+
+        // Check that target is in the range of the NSEC and that owner is a
+        // prefix of the next_name.
+        if nsec_in_range(target, &owner, &nsec.next_name())
+            && nsec.next_name().ends_with(target)
+        {
+            return (NsecState::NoData, None);
+        }
+
+        // No match, try the next one.
+    }
+    (NsecState::Nothing, ede)
+}
+
+// Find an NSEC record for target that proves that the target does not
+// exist. Then find an NSEC record for a wildcard that covers the target
+// and check that nothing for rtype exists.
+//
+// So we have two possibilities: we find an exact match for the wildcard and
+// check the bitmap or we find the wildcard as an empty non-terminal.
+pub fn nsec_for_nodata_wildcard(
+    target: &Name<Bytes>,
+    groups: &mut [ValidatedGroup],
+    rtype: Rtype,
+    signer_name: &Name<Bytes>,
+) -> (NsecState, Option<ExtendedError<Vec<u8>>>) {
+    let (state, ede) = nsec_for_not_exists(target, groups, signer_name);
+    let ce = match state {
+        NsecNXState::DoesNotExist(ce) => ce,
+        NsecNXState::Nothing => return (NsecState::Nothing, ede),
+        NsecNXState::Exists => {
+            // We have proof that the name exists but we are trying to proof
+            // that it doesn't exist. Just pretend that we didn't find proof
+            // that it doesn't exist.
+            return (NsecState::Nothing, ede);
+        }
+    };
+
+    let star_name = match star_closest_encloser(&ce) {
+        Ok(name) => name,
+        Err(_) => {
+            // We cannot create the wildcard name. Pretend that we didn't
+            // find anything and return a suitable ExtendedError.
+            let ede = make_ede(
+                ExtendedErrorCode::DNSSEC_BOGUS,
+                "cannot create wildcard record",
+            );
+            return (NsecState::Nothing, ede);
+        }
+    };
+    nsec_for_nodata(&star_name, groups, rtype, signer_name)
+}
+
+#[derive(Debug)]
+pub enum NsecNXState {
+    Exists,
+    DoesNotExist(Name<Bytes>),
+    Nothing,
+}
+
+// Find an NSEC record for target that proves that the target does not
+// exist. Return the status and the closest encloser.
+pub fn nsec_for_not_exists(
+    target: &Name<Bytes>,
+    groups: &mut [ValidatedGroup],
+    signer_name: &Name<Bytes>,
+) -> (NsecNXState, Option<ExtendedError<Vec<u8>>>) {
+    let mut ede = None;
```

Comment:
I made the following change:
            if ede.is_none() {
                // get_checked_nsec may have provided a reason it could not
                // return an NSEC record. Typically this happens if there is
                // an NSEC record but it fails some validation check. Store
                // the first reason we got to (hopefully) provide a clue to the
                // user if validation fails later on.
                ede = new_ede;
            }